distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Extensions.enso

```diff
@@ -58,6 +59,28 @@ Text.length =
         @Tail_Call count (accum + 1) iterator.next
     count 0 iterator.next
 
+## Returns a new `Text` object with the characters in the reverse order of the input.
+
+   ! What is a Character?
+     A character is defined as an Extended Grapheme Cluster, see Unicode
+     Standard Annex 29. This is the smallest unit that still has semantic
+     meaning in most text-processing applications.
+
+   > Example
+     Reverse the text "Hello, world!".
+
+       "Hello, world!".reverse
+Text.reverse : Text
+Text.reverse =
+    reverseStringBuilder = StringBuilder.new this.length
+    iterator = BreakIterator.getCharacterInstance
+    iterator.setText this
+
+    iterate prev next = if next == -1 then reverseStringBuilder.toString else
+      reverseStringBuilder.append (Text_Utils.substring this next prev)
```

Comment:
Using a `StringBuilder` works, but I'm wondering if using our `+` wouldn't be more idiomatic -  for example our `Vector.join` does in fact use `+`.

Performance-wise it should be quite comparable, because our `+` uses ropes under the hood to achieve O(1) amortized append time. Still, probably `StringBuilder` has a bit less of an overhead, so likely still a bit faster.

So I'm not really saying to change it - because this definitely works and is fast, but wondering what best practices we should adopt - because I think it would be good to have consistent approach to this between this function, `Vector.join` and possibly other ones.

Reply:
`Vector.join` is using memory copying for speed, if we ended up with a vector class working with ropes then would make sense but currently we are ensuring that the underlying data structure is a sequential array.

Additionally, I wonder on our Text ropes are we caching the collapsed string when we fold it. Many of our functions rely on the char[] or string representation. Probably one for optimisation when get to speed hunting!

Reply:
Yes, Text ropes cache collapsed strings - see `ToJavaStringNode::inplaceFlatten`.

As for `Vector.join` - I'm talking about the function which concatenates a vector of texts into a single text, did you mean `Vector.+` instead?

Reply:
@radeusgd ping ^

Reply:
Hmm, then I think you are right that it may be better to keep the stringbuilder. I guess the difference is because the ropes still have some more allocation overhead.

As for `Vector.join` and `Vector.+` - no worry it was a reply to a later deleted comment, please ignore :sweat_smile: